File: app-python/entitlements_app/services/set_account_override_service.py
"""Create an account-specific entitlement override for support or sales operations."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.set_account_override_request import SetAccountOverrideRequest
from entitlements_app.entities import AccountOverride
from entitlements_app.ports import EntitlementsRepository


class SetAccountOverrideService:
    """Service handling Command `SetAccountOverride`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: SetAccountOverrideRequest) -> AccountOverride:
        # spec-driven body for Command SetAccountOverride.
        # intent: Create an account-specific entitlement override for support or sales operations.
        # requires: EntitlementsRepository
        # emits events: AccountOverrideSet
        # return: AccountOverride (entity); construct from request inputs and persist via EntitlementsRepository.saveAccountOverride.
        value = AccountOverride(tenantId=request.tenantId, overrideId=request.overrideId, accountId=request.accountId, featureKey=request.featureKey, overrideType=request.overrideType, allowance=request.allowance, reason=request.reason, expiresAt=request.expiresAt, createdAt=request.createdAt)
        return self.entitlements_repository.saveAccountOverride(value)

File: app/src/main/java/com/example/spring/SetAccountOverrideService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Create an account-specific entitlement override for support or sales operations.
 */
@Service
public class SetAccountOverrideService {

    private final EntitlementsRepository entitlementsRepository;
    private final ApplicationEventPublisher springEventPublisher;

    public SetAccountOverrideService(EntitlementsRepository entitlementsRepository, ApplicationEventPublisher springEventPublisher) {
        this.entitlementsRepository = entitlementsRepository;
        this.springEventPublisher = springEventPublisher;
    }

    @Transactional
    public AccountOverride handle(SetAccountOverrideRequest request) {
        // === GEN_REGION_START: service-body ===
        // spec-driven body for SetAccountOverride.
        // intent: Create an account-specific entitlement override for support or sales operations.
        // requires: EntitlementsRepository
        // emits events: AccountOverrideSet
        // return: AccountOverride (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveAccountOverride.
        AccountOverride value = new AccountOverride(request.tenantId(), request.overrideId(), request.accountId(), request.featureKey(), request.overrideType(), request.allowance(), request.reason(), request.expiresAt(), request.createdAt());
        AccountOverride saved = this.entitlementsRepository.saveAccountOverride(value);
        this.springEventPublisher.publishEvent(new AccountOverrideSet(saved.accountId(), saved.allowance(), saved.createdAt(), saved.featureKey(), saved.overrideId(), saved.overrideType(), saved.tenantId()));
        return saved;
        // === GEN_REGION_END: service-body ===
    }
}